From 8d344e6809f5e2564035f2079a96a19f17a665eb Mon Sep 17 00:00:00 2001
From: Shengjiu Wang <shengjiu.wang@freescale.com>
Date: Thu, 13 Apr 2017 11:13:04 +0800
Subject: [PATCH 1672/5242] MLK-14525: ASoC: fsl_esai: channel swap issue in 3
 channels or 5 channels

commit  8855e5bb28f8318ea6961afe8d449778c91d6bfc from
https://source.codeaurora.org/external/imx/linux-imx.git

In 3 channels, ESAI is set to I2S mode and two dataline, so actually
ESAI is working in 4 channels mode, (we can't get correct 3 channels,
for we can't make one dataline working in 2 channel, another dataline
working in 1 channels).

In this case, we fill channels (3) zero data to FIFO in start phase,
which should cause channel swap, that we need to fill 4 zero data to
FIFO for ESAI working in 4 channels mode actually.

Signed-off-by: Shengjiu Wang <shengjiu.wang@freescale.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 sound/soc/fsl/fsl_esai.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/sound/soc/fsl/fsl_esai.c b/sound/soc/fsl/fsl_esai.c
index 8eadd2d0..f0a76af 100644
--- a/sound/soc/fsl/fsl_esai.c
+++ b/sound/soc/fsl/fsl_esai.c
@@ -3,6 +3,7 @@
 // Freescale ESAI ALSA SoC Digital Audio Interface (DAI) driver
 //
 // Copyright (C) 2014-2016 Freescale Semiconductor, Inc.
+// Copyright 2017 NXP
 
 #include <linux/clk.h>
 #include <linux/dmaengine.h>
@@ -604,7 +605,7 @@ static int fsl_esai_trigger(struct snd_pcm_substream *substream, int cmd,
 				   ESAI_xFCR_xFEN_MASK, ESAI_xFCR_xFEN);
 
 		/* Write initial words reqiured by ESAI as normal procedure */
-		for (i = 0; tx && i < channels; i++)
+		for (i = 0; tx && i < pins * esai_priv->slots; i++)
 			regmap_write(esai_priv->regmap, REG_ESAI_ETDR, 0x0);
 
 		regmap_update_bits(esai_priv->regmap, REG_ESAI_xCR(tx),
-- 
1.7.9.5

